File: Cpp/445_Add_Two_Numbers_II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* result = nullptr;
        stack<ListNode*> s1;
        stack<ListNode*> s2;
        ListNode* current = l1;
        while(current != nullptr) {
            s1.push(current);
            current = current->next;
        }
        current = l2;
        while(current != nullptr) {
            s2.push(current);
            current = current->next;
        }
        
        if(s1.size() < s2.size()) std::swap(s1, s2);
        current = nullptr;
        int carry = 0;
        int num1, num2;
        while(!s1.empty()) {
            if(!s1.empty()) {
                num1 = s1.top()->val;
                current = s1.top();
                s1.pop();
            } else {
                num1 = 0;
            }

            if(!s2.empty()) {
                num2 = s2.top()->val;
                s2.pop();
            } else {
                num2 = 0;
            }

            int sum = num1 + num2 + carry;
            carry = sum / 10;
            current->val = sum % 10;
            
            current->next = result;
            result = current;
        }
        
        if(carry) {
            current = new ListNode(1);
            current->next = result;
            result = current;
        }
        
        return result;
    }
};

File: Kotlin/445_Add_Two_Numbers_II.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun addTwoNumbers(l1: ListNode?, l2: ListNode?): ListNode? {
        var result : ListNode? = null
        if(l1 == null) return l2
        if(l2 == null) return l1
        
        var s1 = Stack<ListNode>()
        var s2 = Stack<ListNode>()
        
        var current = l1
        while (current != null) {
            s1.push(current)
            current = current.next
        }
        
        current = l2
        while (current != null) {
            s2.push(current)
            current = current.next
        }
        
        if(s1.size <s2.size) {
            val s3 = s2
            s2 = s1
            s1 = s3
        }
        current = null
        var carry = 0
        while (!s1.empty()) {
            var sum = carry
            if(!s1.empty()) {
                current = s1.pop()
                sum += current.`val`
            }
            if(!s2.empty()) {
                sum += s2.pop().`val`
            }
            
            carry = sum / 10
            current?.`val` = sum % 10
            
            current?.next = result
            result = current
        }
        
        if(carry > 0) {
            current = ListNode(1)
            current.next = result
            result = current
        }
        return result
    }
}